### DISCUSSION TITLE
How to get off-diagonal Green's function matrix elements?

### DISCUSSION THREAD

**Discussion by chillenb** [block-hczhai/block2-preview#101] [Q&A] [answered]
I'm following the example [here](https://block2.readthedocs.io/en/latest/tutorial/greens-function.html) and would like to get off-diagonal elements of the Green's function. The example suggests the use of `driver.expectation()`. Unfortunately, I haven't been able to get it to work:

```
In [3]: driver.greens_function(bra, mpo, dmpo, ket, 0, eta, n_sweeps=6, bra_bond_dims=[200], ket_bond_dims=[200], thrds=[1E-6] * 10, iprint=0)
(0.9686338401489134-0.010675119181648821j)

In [4]: driver.expectation(bra, driver.get_identity_mpo(), dket)
-0.010675119181648438

In [5]: driver.expectation(bra, dmpo, ket)
-0.0106751191816489
```

It looks like `driver.expectation()` is only giving the imaginary component of the inner product because block2 is in real mode. I've tried running the example in complex mode, but it throws an error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[48], line 41
     39 for iw, freq in enumerate(freqs):
     40     bra = driver.copy_mps(dket, tag="BRA") # initial guess
---> 41     gfmat[iw] = driver.greens_function(bra, mpo, dmpo, ket, freq, eta, n_sweeps=6,
     42         bra_bond_dims=[200], ket_bond_dims=[200], thrds=[1E-6] * 10, iprint=0)
     43     print("FREQ = %8.2f GF[%d,%d] = %12.6f + %12.6f i" % (freq, isite, isite, gfmat[iw].real, gfmat[iw].imag))
     45 ldos = -2 [/](http://pizza:8888/) np.pi * gfmat.imag # account for both spin

File [~/miniforge3/envs/block2prod/lib/python3.11/site-packages/pyblock2/driver/core.py:4021](http://pizza:8888/home/chillenb/miniforge3/envs/block2prod/lib/python3.11/site-packages/pyblock2/driver/core.py#line=4020), in DMRGDriver.greens_function(self, bra, mpo, rmpo, ket, omega, eta, n_sweeps, tol, bra_bond_dims, ket_bond_dims, noises, thrds, cutoff, linear_max_iter, iprint)
   4018 cps.gf_omega = omega
   4020 cps.solve(n_sweeps, ket.center == 0, tol)
-> 4021 rgf, igf = cps.targets[-1]
   4023 if self.clean_scratch:
   4024     lme.remove_partition_files()

ValueError: not enough values to unpack (expected 2, got 1)
```

If somebody could show me the right way to do this, I'd be very grateful!

**Comment by hczhai** [comment, marked as answer]
The complex mode problem is fixed in https://github.com/block-hczhai/block2-preview/commit/a5e1e84f1388e80fff43a03f9e6960937e97b94b.

For the real mode, you may need to apply H to the imag part to get the real part MPS (using ``driver.multiply``, for example), and then compute the expectations.

If you want to do these more efficiently (without doing multiple sweeps), you need to use the low-level Python interface like what is done in https://github.com/ZhuGroup-Yale/fcdmft/blob/main/fcdmft/solver/gfdmrg.py#L952-L1038. This highly optimized procedure is too complicated so currently it is not fully supported in the ``DMRGDriver`` high-level interface.

**Reply by chillenb** [reply]
Got it! Thank you so much for the quick response.
I will try both methods, particularly the one from fcdmft. That one has everything I need and looks like it is worth studying in detail to become more familiar with block2.
